### PR TITLE
Throw exception when contain() in query builder for matching

### LINKED ISSUE #12440: Query error when using contain() inside innerJoinWith()
This is a (multiple allowed):

* [x] bug
* [ ] enhancement
* [ ] feature-discussion (RFC)

* CakePHP Version: 3.5.17

I've found a problem using the ShadowTranslate behavior ([AD7six/cakephp-shadow-translate](url)).

Please, see also [AD7six/cakephp-shadow-translate/issues/52](url) for more information.

I have the associations: Destinations hasMany Countries hasMany CountriesI18n

This is an example that shows what the behavior is trying to do:

```php
$DestinationsTable
  ->find()
  ->innerJoinWith('Countries', function ($q) {
    return $q->contain('CountriesI18n');
   })
  ->where(['CountriesI18n.slug' => 'aaa']);
```
The result:
> PDOException with message 'SQLSTATE[42S22]: Column not found: 1054 Unknown column 'CountriesI18n.slug' in 'where clause''

This is the resulting query:

```
=> Cake\ORM\Query {#334                     
     +"(help)": "This is a Query object, to get the results execute or iterate it.",
     +"sql": "SELECT Destinations.id AS `Destinations__id`, Destinations.created AS
`Destinations__created`, Destinations.type AS `Destinations__type`,
Destinations.province_id AS `Destinations__province_id`,
Destinations.country_id AS `Destinations__country_id`,
Destinations.parent_id AS `Destinations__parent_id`, Destinations.enabled AS
`Destinations__enabled`, Destinations.name AS `Destinations__name`
FROM destinations Destinations
INNER JOIN countries Countries ON Countries.id = (Destinations.country_id)
WHERE CountriesI18n.slug = :c0",
     +"params": [                          
       ":c0" => [             
         "value" => "aaa",                  
         "type" => null,                   
         "placeholder" => "c0",
       ],                                   
     ],                                    
     +"defaultTypes": [
....
   ],
     +"decorators": 0,
     +"executed": false,
     +"hydrate": true,
     +"buffered": true,
     +"formatters": 0,
     +"mapReducers": 0,
     +"contain": [
       "Countries" => [
         "CountriesI18n" => [],
       ],
     ],
     +"matching": [
       "Countries" => [
         "queryBuilder" => Closure {#348
           parameters: {
             $q: {},
           },
         },
         "matching" => true,
         "joinType" => "INNER",
         "fields" => false,
       ],
     ],
     +"extraOptions": [],
     +"repository": App\Model\Table\DestinationsTable {#301
       +"registryAlias": "Destinations",
       +"table": "destinations",
       +"alias": "Destinations",
       +"entityClass": "App\Model\Entity\Destination",
       +"associations": [
         "countries",
       ],
       +"behaviors": [],
       +"defaultConnection": "default",
       +"connectionName": "default",
     },
   }
```

I'm trying to understand the issue and fix it but I'm not sure where and how to fix it.

Thanks.

### PR DESCRIPTION
**Pull request by othercorey**
Fixes https://github.com/cakephp/cakephp/issues/12440

Original test case
https://github.com/cakephp/cakephp/pull/12693

The original issue doesn't describe the failure properly because `hasMany` will load through select.

This tests against the `matching` option set by `matching()` or `innerJoinWith()` that gets passed to the association from the eager loader although it's not one of the options consumed by `attachTo()`.

### CODE REVIEW

File: src/ORM/Association.php [hunk 1]
```diff
@@ -749,6 +749,15 @@ public function attachTo(Query $query, array $options = []): void
                     $this->getName()
                 ));
             }
+            if (
+                !empty($options['matching']) &&
+                $this->_strategy === static::STRATEGY_JOIN &&
+                !empty($dummy->getContain())
```
Comment (dereuromark) [inline src/ORM/Association.php:755]: there should be no method usage here in empty() calls, it can be removed without issues here.

File: tests/TestCase/ORM/QueryTest.php [hunk 2]
```diff
@@ -3576,6 +3578,29 @@ public function testInnerJoinWithSelect()
...
+    {
+        $comments = $this->getTableLocator()->get('Comments');
+        $articles = $comments->belongsTo('Articles');
+        $articles->hasOne('ArticlesTranslations');
+
+        $this->expectException(RuntimeException::class);
+        $this->expectExceptionMessage('Cannot contain associations in query returned by query builder');
+        $comments
+            ->find()
+            ->innerJoinWith('Articles', function ($q) {
```
Comment (dereuromark) [inline tests/TestCase/ORM/QueryTest.php:3596]: Can you please typehint $q here, as the ->contain() and ->where() are not detected right now (not followable).

File: src/ORM/Association.php (hunk 1 above)
Comment (othercorey) [reply src/ORM/Association.php:755]: Can you rephrase that?

Comment (dereuromark) [reply src/ORM/Association.php:755]: `&& $dummy->getContain()`
suffices

The cloaking method check is not needed and an anti pattern in these cases of existing property, key or method.

Comment (othercorey) [reply src/ORM/Association.php:755]: That's fine, just couldn't understand the sentence so wanted to double check.

File: src/ORM/Association.php [hunk 3]
```diff
@@ -749,6 +749,15 @@ public function attachTo(Query $query, array $options = []): void
                     $this->getName()
                 ));
             }
+            if (
+                !empty($options['matching']) &&
+                $this->_strategy === static::STRATEGY_JOIN &&
+                $dummy->getContain()
+            ) {
+                throw new RuntimeException(
+                    "Cannot contain associations in query returned by query builder for `{$this->getName()}`."
```
Comment (markstory) [inline src/ORM/Association.php:758]:
```suggestion
                    "Cannot contain associations in queries returned by query builder Closures for `{$this->getName()}`."
```

File: tests/TestCase/ORM/QueryTest.php [hunk 4]
```diff
@@ -3576,6 +3578,28 @@ public function testInnerJoinWithSelect()
...
+     * @return void
+     */
+    public function testInnerJoinWithContain()
+    {
+        $comments = $this->getTableLocator()->get('Comments');
+        $articles = $comments->belongsTo('Articles');
+        $articles->hasOne('ArticlesTranslations');
+
+        $this->expectException(RuntimeException::class);
+        $this->expectExceptionMessage('Cannot contain associations in query returned by query builder');
```
Comment (markstory) [inline tests/TestCase/ORM/QueryTest.php:3593]:
```suggestion
        $this->expectExceptionMessage('Cannot contain associations in queries returned by query builder');
```

File: src/ORM/Association.php (hunk 3 above)
Comment (othercorey) [reply src/ORM/Association.php:758]: Should we fix the other exception messages to use `Closure`?

`'Query builder for association "%s" did not return a query'`

Comment (othercorey) [reply src/ORM/Association.php]: Never mind -- closure isn't really relevant. I adjusted the test to happen no matter how the contain is set.
